#ID3 Algorithm (for Classification)
ID3 (Iterative Dichotomiser 3) is used for building classification trees, primarily for categorical features. It uses information gain to select the splitting attribute.
<br><br>
**Start with the Entire Dataset**: The root node of the tree represents the entire training dataset.<br>
**Calculate Impurity (Entropy):** For the current node, calculate the entropy of the target variable. This measures the disorder or uncertainty in the dataset.<br>
**Evaluate Attributes:** For each available feature (attribute):
Calculate the entropy for each possible value of the feature.
Calculate the information gain achieved by splitting on that feature. Information gain is the reduction in entropy after splitting.<br>
**Select Best Attribute:** Choose the feature that yields the highest information gain. This is the attribute that best separates the data according to the target variable.<br>
**Create Child Nodes:** Create child nodes for each unique value of the chosen best attribute.<br>
**Split Dataset:** Partition the dataset into subsets, one for each child node, based on the values of the selected attribute.<br>
**Recurse:** Recursively apply steps 2-6 to each child node, using only the remaining attributes.<br>
**Stop Condition:** The recursion stops when:<br>
  All instances in a node belong to the same class (the node is 'pure').<br>
  There are no remaining attributes to split on.<br>
  There are no instances left in the node.<br>
**Assign Leaf Node:** When a stop condition is met, the node becomes a leaf node and is labeled with the majority class of the instances it contains.<br><br><br>
#CART Algorithm (for Regression)
CART (Classification and Regression Trees) can be used for both classification and regression. When used for regression, it aims to minimize the variance (or squared error) within each node.<br><br>

**Start with the Entire Dataset:** The root node represents the entire training dataset. The initial prediction for this node is the mean of the target variable for all samples.<br>
**Calculate Impurity (Squared Error/Variance):** For the current node, calculate the total squared error (or variance) of the target variable. This measures how spread out the target values are.<br>
**Evaluate Possible Splits:** For each feature and for each possible split point (value) within that feature:<br>
  Imagine splitting the node into two child nodes based on that split point.<br>
  Calculate the squared error for each potential child node.<br>
  Calculate the reduction in overall squared error (or variance reduction) achieved by that split.<br>
**Select Best Split:** Choose the feature and split point that results in the greatest reduction in squared error (i.e., creates the most homogeneous child nodes).<br>
**Create Child Nodes:** Create two child nodes based on the best split, one for instances less than or equal to the split point, and one for instances greater than the split point.<br>
**Split Dataset:** Partition the dataset into two subsets corresponding to the child nodes.<br>
Recurse: Recursively apply steps 2-6 to each child node.<br>
**Stop Condition:** The recursion stops when:<br>
A node contains a number of instances below a predefined minimum (e.g., min_samples_split).<br>
The reduction in squared error for any potential split falls below a certain threshold.<br>
The maximum tree depth is reached.<br>
**Assign Leaf Node:** When a stop condition is met, the node becomes a
leaf node. The prediction for this leaf node is the mean of the target variable of all instances it contains.<br>

## Generate Classification Dataset

Generate a synthetic dataset suitable for binary classification with categorical features, to be used for the ID3 algorithm.


In [ ]:
import numpy as np
import pandas as pd

# Create a dictionary with categorical features and a binary target variable
data = {
    'Feature1': np.random.choice(['Red', 'Green', 'Blue'], size=100),
    'Feature2': np.random.choice(['Small', 'Medium', 'Large'], size=100),
    'Feature3': np.random.choice(['High', 'Low'], size=100),
    'Target': np.random.choice([0, 1], size=100) # Binary target variable
}

# Create a pandas DataFrame from the dictionary
df = pd.DataFrame(data)

# Display the first few rows of the DataFrame
print("Generated DataFrame:")
print(df.head())

Generated DataFrame:
  Feature1 Feature2 Feature3  Target
0    Green   Medium      Low       0
1      Red    Large     High       0
2      Red   Medium     High       1
3     Blue    Large     High       0
4    Green    Large     High       1


## Build and Train ID3 Classifier

Build and train an ID3-like decision tree classifier (using entropy criterion for impurity) on the generated classification dataset.


In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Separate features (X) and target (y)
X = df[['Feature1', 'Feature2', 'Feature3']]
y = df['Target']

# Convert categorical features to numerical using one-hot encoding
X_encoded = pd.get_dummies(X)

# Instantiate DecisionTreeClassifier with entropy criterion
id3_classifier = DecisionTreeClassifier(criterion='entropy', random_state=42)

# Train the decision tree classifier
id3_classifier.fit(X_encoded, y)

print("Features (X) head after one-hot encoding:")
print(X_encoded.head())
print("\nDecision Tree Classifier trained successfully.")

Features (X) head after one-hot encoding:
   Feature1_Blue  Feature1_Green  Feature1_Red  Feature2_Large  \
0          False            True         False           False   
1          False           False          True            True   
2          False           False          True           False   
3           True           False         False            True   
4          False            True         False            True   

   Feature2_Medium  Feature2_Small  Feature3_High  Feature3_Low  
0             True           False          False          True  
1            False           False           True         False  
2             True           False           True         False  
3            False           False           True         False  
4            False           False           True         False  

Decision Tree Classifier trained successfully.


## Classify New Sample with ID3

Use the trained ID3 classifier to classify a new, unseen data sample and display the predicted class.


In [ ]:
import pandas as pd

# 1. Create a new pandas DataFrame for a new, unseen sample
new_sample_df = pd.DataFrame({
    'Feature1': ['Blue'],
    'Feature2': ['Small'],
    'Feature3': ['Low']
})

# 2. Apply one-hot encoding to the new sample and align columns with X_encoded
new_sample_encoded = pd.get_dummies(new_sample_df).reindex(columns=X_encoded.columns, fill_value=0)

# 3. Use the trained id3_classifier to predict the class of the new sample
predicted_class = id3_classifier.predict(new_sample_encoded)

# 4. Print the new sample and the predicted class
print("New Sample to Classify:")
print(new_sample_df)
print(f"\nPredicted Class for the new sample: {predicted_class[0]}")

New Sample to Classify:
  Feature1 Feature2 Feature3
0     Blue    Small      Low

Predicted Class for the new sample: 1


## Generate Regression Dataset

Generate a synthetic dataset suitable for regression with numerical features, to be used for the CART algorithm.

In [ ]:
import numpy as np
import pandas as pd

# Create a dictionary with numerical features and a continuous target variable
data_reg = {
    'FeatureA': np.random.rand(100) * 100,  # Continuous numerical feature
    'FeatureB': np.random.rand(100) * 50,   # Continuous numerical feature
    'FeatureC': np.random.randint(0, 10, size=100), # Integer numerical feature
    'Target_Reg': (np.random.rand(100) * 50) + (np.random.rand(100) * 20) + (np.random.rand(100) * 30) # Continuous target variable
}

# Create a pandas DataFrame from the dictionary
df_reg = pd.DataFrame(data_reg)

# Display the first few rows of the DataFrame
print("Generated Regression DataFrame:")
print(df_reg.head())

Generated Regression DataFrame:
    FeatureA   FeatureB  FeatureC  Target_Reg
0  57.984595  40.783663         2   28.748123
1  94.681795  42.372396         3   51.505217
2  37.040641  49.614260         0   25.912798
3  72.214747   5.643247         6   21.596316
4  81.243762  18.084546         3   74.537566


## Build and Train CART Regressor


**Reasoning**:
To build and train the CART decision tree regressor, first separate the features from the target variable in `df_reg`. Then, instantiate a `DecisionTreeRegressor` with the squared error criterion and train it on the prepared data.



In [ ]:
from sklearn.tree import DecisionTreeRegressor

# Separate features (X_reg) and target (y_reg)
X_reg = df_reg[['FeatureA', 'FeatureB', 'FeatureC']]
y_reg = df_reg['Target_Reg']

# Instantiate DecisionTreeRegressor with squared error criterion (default)
cart_regressor = DecisionTreeRegressor(criterion='squared_error', random_state=42)

# Train the decision tree regressor
cart_regressor.fit(X_reg, y_reg)

print("Regression features (X_reg) head:")
print(X_reg.head())
print("\nDecision Tree Regressor trained successfully.")

Regression features (X_reg) head:
    FeatureA   FeatureB  FeatureC
0  57.984595  40.783663         2
1  94.681795  42.372396         3
2  37.040641  49.614260         0
3  72.214747   5.643247         6
4  81.243762  18.084546         3

Decision Tree Regressor trained successfully.


## Predict New Sample with CART Regressor

Use the trained CART regressor to predict the target value for a new, unseen data sample and display the predicted value.

In [ ]:
import pandas as pd

# 1. Create a new pandas DataFrame for a new, unseen sample for regression
new_sample_reg_df = pd.DataFrame({
    'FeatureA': [75.0],  # Example numerical value
    'FeatureB': [30.0],  # Example numerical value
    'FeatureC': [5]      # Example integer numerical value
})

# 2. Use the trained cart_regressor to predict the target value of the new sample
predicted_value_reg = cart_regressor.predict(new_sample_reg_df)

# 3. Print the new sample and the predicted target value
print("New Regression Sample to Predict:")
print(new_sample_reg_df)
print(f"\nPredicted Target Value for the new sample: {predicted_value_reg[0]:.2f}")

New Regression Sample to Predict:
   FeatureA  FeatureB  FeatureC
0      75.0      30.0         5

Predicted Target Value for the new sample: 74.81


## Summary:

### Q&A
The ID3 classification model predicted a class of `1` for the new sample `{'Feature1': ['Blue'], 'Feature2': ['Small'], 'Feature3': ['Low']}`. The CART regression model predicted a target value of `74.81` for the new sample `{'FeatureA': [75.0], 'FeatureB': [30.0], 'FeatureC': [5]}`.

### Data Analysis Key Findings
*   A synthetic dataset for binary classification was generated, featuring three categorical variables and a binary target.
*   An ID3-like Decision Tree Classifier, utilizing the entropy criterion, was successfully built and trained on the one-hot encoded classification dataset.
*   The trained ID3 classifier predicted a class of `1` for a new sample with features 'Blue', 'Small', and 'Low'.
*   A synthetic dataset for regression was generated, comprising three numerical features and a continuous target variable.
*   A CART Decision Tree Regressor, using the squared error criterion, was successfully built and trained on the regression dataset.
*   The trained CART regressor predicted a target value of `74.81` for a new regression sample with features `75.0`, `30.0`, and `5`.

### Insights or Next Steps
*   Both classification and regression models successfully made predictions on new, unseen data, indicating their immediate utility for inference tasks.
*   Future work could involve evaluating the performance of these models using appropriate metrics (e.g., accuracy, precision, recall for classification; R-squared, MSE for regression) on dedicated test sets, and exploring hyperparameter tuning to optimize their generalization capabilities.


In [ ]:
# ID3-style Decision Tree Classification using Iris dataset

from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Load builtin dataset
iris = load_iris()

X = iris.data        # features
y = iris.target      # labels

print("Dataset used: IRIS")
print("Feature names:", iris.feature_names)
print("Target classes:", iris.target_names)

# Split data into training and testing sets
X_train_iris, X_test_iris, y_train_iris, y_test_iris = train_test_split(X, y, test_size=0.3, random_state=42)

# Train decision tree using entropy (ID3 logic)
id3_model = DecisionTreeClassifier(criterion="entropy", random_state=42)
id3_model.fit(X_train_iris, y_train_iris)

# Make predictions on the test set
y_pred_iris = id3_model.predict(X_test_iris)

# Calculate and print classification metrics
accuracy = accuracy_score(y_test_iris, y_pred_iris)
precision = precision_score(y_test_iris, y_pred_iris, average='weighted')
recall = recall_score(y_test_iris, y_pred_iris, average='weighted')
f1 = f1_score(y_test_iris, y_pred_iris, average='weighted')

print("\n--- ID3 Classification Model Metrics ---")
print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print(f"F1-Score: {f1:.2f}")

# New sample for prediction (using the original full dataset's model for consistency with the initial prompt)
sample = [[5.1, 3.5, 1.4, 0.2]]

# Retrain model on full dataset for the single sample prediction as per original request, if needed
# Or just use the test model if the new sample is part of the test set
# For simplicity, will re-use the trained id3_model from above for the new sample
prediction = id3_model.predict(sample)

print("\nNew sample for classification:", sample)
print("Predicted class:", iris.target_names[prediction][0])


Dataset used: IRIS
Feature names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Target classes: ['setosa' 'versicolor' 'virginica']

--- ID3 Classification Model Metrics ---
Accuracy: 0.98
Precision: 0.98
Recall: 0.98
F1-Score: 0.98

New sample for classification: [[5.1, 3.5, 1.4, 0.2]]
Predicted class: setosa


In [ ]:
# CART Regression using Diabetes dataset

from sklearn.datasets import load_diabetes
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Load builtin dataset
data = load_diabetes()

X_diabetes = data.data
y_diabetes = data.target

print("Dataset used: DIABETES")
print("Number of features:", X_diabetes.shape[1])

# Split data into training and testing sets
X_train_diabetes, X_test_diabetes, y_train_diabetes, y_test_diabetes = train_test_split(X_diabetes, y_diabetes, test_size=0.3, random_state=42)

# Train regression tree
cart_model = DecisionTreeRegressor(random_state=42)
cart_model.fit(X_train_diabetes, y_train_diabetes)

# Make predictions on the test set
y_pred_diabetes = cart_model.predict(X_test_diabetes)

# Calculate and print regression metrics
mse = mean_squared_error(y_test_diabetes, y_pred_diabetes)
r2 = r2_score(y_test_diabetes, y_pred_diabetes)

print("\n--- CART Regression Model Metrics ---")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared: {r2:.2f}")

# New sample (first record used as example)
sample = [X_diabetes[0]]

prediction = cart_model.predict(sample)

print("\nNew sample for regression:", sample[0])
print("Predicted disease progression:", prediction[0])


Dataset used: DIABETES
Number of features: 10

--- CART Regression Model Metrics ---
Mean Squared Error (MSE): 5697.79
R-squared: -0.06

New sample for regression: [ 0.03807591  0.05068012  0.06169621  0.02187239 -0.0442235  -0.03482076
 -0.04340085 -0.00259226  0.01990749 -0.01764613]
Predicted disease progression: 293.0
